src/MongoDB.Driver.Core/Core/WireProtocol/Messages/CompressedMessage.cs

```diff
@@ -0,0 +1,75 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using MongoDB.Bson.IO;
+using MongoDB.Driver.Core.Compression;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.WireProtocol.Messages.Encoders;
+
+namespace MongoDB.Driver.Core.WireProtocol.Messages
+{
+    /// <summary>
+    /// Represents a compressed message.
+    /// </summary>
+    public class CompressedMessage : MongoDBMessage
+    {
+        private readonly BsonStream _originalMessageBytes;
+        private readonly MongoDBMessage _originalMessage;
+        private readonly CompressorType _compressorType;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompressedMessage"/> class.
+        /// </summary>
+        /// <param name="originalMessage">The original message.</param>
+        /// <param name="originalBytes">The original bytes.</param>
+        /// <param name="compressorType">The compressor type.</param>
+        public CompressedMessage(
+            MongoDBMessage originalMessage,
+            BsonStream originalBytes,
+            CompressorType compressorType)
+        {
+            _originalMessage = Ensure.IsNotNull(originalMessage, nameof(originalMessage));
+            _compressorType = compressorType;
+            _originalMessageBytes = originalBytes;
+        }
```

Comment:
I think we need two constructors, one used when writing a compressed message and another used when reading one:

        /// <summary>
        /// Initializes a new instance of the <see cref="CompressedMessage"/> class.
        /// This constructor is used when reading a compressed message.
        /// </summary>
        /// <param name="originalMessage">The original message.</param>
        public CompressedMessage(MongoDBMessage originalMessage)
        {
            _originalMessage = Ensure.IsNotNull(originalMessage, nameof(originalMessage));
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="CompressedMessage"/> class.
        /// This constructor is used when writing a compressed message.
        /// </summary>
        /// <param name="originalMessage">The original message.</param>
        /// <param name="originalMessageStream">The original message bytes.</param>
        /// <param name="compressorType">The compressor type.</param>
        public CompressedMessage(
            MongoDBMessage originalMessage,
            BsonStream originalMessageStream,
            CompressorType compressorType)
        {
            _originalMessage = Ensure.IsNotNull(originalMessage, nameof(originalMessage));
            _originalMessageStream = originalMessageStream;
            _compressorType = compressorType;
        }